Dtorresf/ch01s01 (#3)
RedHatQuickCourses/rhods-deploy: modules/chapter1/pages/section1.adoc

Reviewer: There's a number of files that we instruct the student to create by copying blocks of code from the instructions. How about we ask them to clone a repository in which these files have been created already? We can then refer to these files and get the student used to a git-based workflow within the workbench as well as for app deployment on OpenShift. We're using this approach in other parts of the course already.

Author: refactored in commit 3408995c7398e110fdb3803941dfb245f275b6cc, waiting for PR https://github.com/RedHatQuickCourses/rhods-qc-apps/pull/12 to allow code in rhods-qc-apps repo.

= Section 1

== The model

A machine learning model is a program that can learn and make predictions or decisions based on data. It's a fundamental component of machine learning, a subset of artificial intelligence. Machine learning models are designed to automatically improve their performance on a task through experience (i.e., training data) without being explicitly programmed for that task.

For example, in natural language processing, machine learning models can parse and correctly recognize the intent behind previously unheard sentences or combinations of words. In image recognition, a machine learning model can be taught to recognize objects - such as cars or dogs. A machine learning model can perform such tasks by having it __'trained'__ with a large dataset. During training, the machine learning algorithm is optimized to find certain patterns or outputs from the dataset, depending on the task. The output of this process - often a computer program with specific rules and data structures - is called a machine learning model.

> The process of creating a model include steps for data collection and selection, the selection of an algorithm to process the data, training the model with such algorithm by tuning its hyperparameters, and once it is measured to produce the required results, exporting the model so that it can be deployed.

Author: contents do not completely align with the document:
After describing the unit 1/section 1 in the [Design Doc](https://docs.google.com/document/d/1PLZ2NrFOWDWyxTW7ei68khZqOMd_GYOnfLFQG4dGyGU/edit#heading=h.v8fdfu8e5zsj), I realize that I would like to draw a history line between some of the methods used to serve models in automated fashion, using applications and images that import models, and working our way up to using the OpenShift.AI tools to make such process easier. This first section is manual, and tries to resolve the problem in an "old fashion" way. In the following sections I will cover how the tools in OpenShift.AI makes the process easier, and as we go using some of the concepts, I will be explaining them.

> The process of creating a model include steps for data collection and selection, the selection of an algorithm to process the data, training the model with such algorithm by tuning its hyperparameters, and once it is measured to produce the required results, exporting the model so that it can be deployed.

Reviewer 2: I'd suggest you double check this with @rsriniva cause the main objective of this section is:

* Describe the concepts and terminology around model serving on RHODS

We can keep the "old-style" exercises to build that history line, that's a great idea actually.
However, we are still missing [some of the terminology concepts in this section](https://docs.google.com/document/d/1PLZ2NrFOWDWyxTW7ei68khZqOMd_GYOnfLFQG4dGyGU/edit#heading=h.v8fdfu8e5zsj) (e.g. what's a model server). Again, double check with @rsriniva .

image::ml_workflow.drawio.svg[Machine Learning Workflow]

== Train a model

Using a RHODS instance, let us train and deploy an example.

> . Open a workbench in RHODS, here you have the option to use the **JupyterLab application**, or to create a **Data Science Project** + **Workbench**

Author: This also requires a change in the picture shown, because instead of depicting both options, should be showing the workbench only.

= Section 1

== The model

A machine learning model is a program that can learn and make predictions or decisions based on data. It's a fundamental component of machine learning, a subset of artificial intelligence. Machine learning models are designed to automatically improve their performance on a task through experience (i.e., training data) without being explicitly programmed for that task.

For example, in natural language processing, machine learning models can parse and correctly recognize the intent behind previously unheard sentences or combinations of words. In image recognition, a machine learning model can be taught to recognize objects - such as cars or dogs. A machine learning model can perform such tasks by having it __trained__ with a large dataset. During training, the machine learning algorithm is optimized to find certain patterns or outputs from the dataset, depending on the task. The output of this process - often a computer program with specific rules and data structures - is called a machine learning model.

> The process of creating a model include steps for data collection and selection, the selection of an algorithm to process the data, training the model with such algorithm by tuning its hyperparameters, and once it is measured to produce the required results, exporting the model so that it can be deployed.

Author: this is why I used to keep the source of these diagrams ;) I'm working on trying to restore the file now.

> The process of creating a model include steps for data collection and selection, the selection of an algorithm to process the data, training the model with such algorithm by tuning its hyperparameters, and once it is measured to produce the required results, exporting the model so that it can be deployed.

Author: refactor of a future comment further modifies this paragraph, see commit b4f0b34, please review.

> The process of creating a model include steps for data collection and selection, the selection of an algorithm to process the data, training the model with such algorithm by tuning its hyperparameters, and once it is measured to produce the required results, exporting the model so that it can be deployed.

Author: refactoring with b4f0b34

* Protobuf

* MLeap

* H5

* ONNX

* PMML

* Torch

The use of either of those formats depend on the target server runtime, some of them are proven to be more eficient than others for certain type of training algorithms and model sizes.
====

== Use the model in another notebook

The model can be deserialized in another notebook, and used to generate a prediction:

. Create a notebook called **use-purchase-amount**
+
image::use-purchase-amount-notebook.png[use-purchase-amount notebook create]

. Use the following code to instantiate the model and request for a prediction
+
```python
# Include required packages for execution
import pickle

# Load models
with open('mymodel.pkl', 'rb') as f:
    model = pickle.load(f)
  
# perform a prediction using the model
print(model(4))
```
+
image::prediction-with-model.png[prediction with model]

[TIP]
====
At this moment the model can be exported and imported in other projects for its use. Normally there will be an S3 bucket or a model registry to store models and versions of such models, and instead of manually exporting the model, there would be pipelines making the model available.
====

> == Use the Model in a Container

Reviewer: This requires the student to install additional dependencies on their laptop, which is not always feasible. Since the laptop is only used for building the container that is then deployed on OCP, how about this workflow instead:
- All files in this section are stored in a github repository.
- The students import this repository into OCP as a new application that includes building the container image based on the Containerfile.
- The students can test the deployed model service as you outlined.

This has the benefit of making the student familiar with a git-based app deployment workflow.

ENTRYPOINT ["gunicorn", "-b", "0.0.0.0:8000", "--access-logfile", "-", "--error-logfile", "-", "--timeout", "120"]
CMD ["app:app"]
```
<1> The python application source code
<2> The list of packages to install
<3> The model

. Build and push the image to an image registry
+
```shell
podman login quay.io
podman build -t purchase-predictor:1.0 .
podman tag purchase-predictor:1.0 quay.io/user_name/purchase-predictor:1.0
podman push quay.io/user_name/purchase-predictor:1.0
```

. Deploy the model image to **OpenShift**
+
```shell
oc login api.cluster.example.com:6443
oc new-project model-deploy
oc new-app --name purchase-predictor quay.io/user_name/purchase-predictor:1.0
oc expose service purchase-predictor
```

Now we can use the Flask application with some commands such as:
```shell
curl http://purchase-predictor-model-deploy.apps.cluster.example.com/health
ok%
curl http://purchase-predictor-model-deploy.apps.cluster.example.com/info
{"name":"Time to purchase amount predictor","version":"v1.0.0"}
curl -d '{"time":4}' -H "Content-Type: application/json" -X POST http://purchase-predictor-model-deploy.apps.cluster.example.com/predict
{"prediction":34,"status":200}
```

[IMPORTANT]
====

> In this section we have manually:

Reviewer: ```suggestion
In this section we have:
```

> 

Reviewer: ```suggestion
. Trained a simple purchase prediction model
```

. Developed an application that uses the model

> . Built an image with such application

Reviewer: ```suggestion
. Built a container image including this application
```

> . Push the image to a registry

Reviewer: ```suggestion
. Pushed the image to a container registry
```

> . Use the image in an OpenShift cluster to create an application

Reviewer: ```suggestion
. Deployed the containerized application in OpenShift
```

> . Expose the application's route in OpenShift

Reviewer: ```suggestion
. Exposed the application's endpoint in OpenShift by creating a route
```

> . Use the application through a REST API to request a prediction

Reviewer: ```suggestion
. Consumed the model through the application's REST API to request a prediction
```

> There are automated and faster ways to perform these steps, in the following sections, we will learn about runtimes that only require you to provide a model, and they automatically resolve provisioning an inference service for you.

Reviewer: ```suggestion
There are automated and faster ways to perform these steps. In the following sections, we will learn about runtimes that only require you to provide a model, and they automatically provision an inference service for you.
```
